### DISCUSSION TITLE
how to cut a texture

### DISCUSSION THREAD

**Discussion by Winlchance** [MonoGame/MonoGame#8415] [Graphics]
I want to write a method to cut texture in order to cut texture atlas, but there is some problems
The code is following:
            public static Texture2D TextureCut(Texture2D texture, int x, int y, int width, int height)
            {
                var cuttedTexture = new Texture2D(texture.GraphicsDevice, width, height);

                Microsoft.Xna.Framework.Color[] data = new Microsoft.Xna.Framework.Color[width * height];
                texture.GetData(0, new Rectangle(x, y, width, height), data, 0, data.Length);
                //
                cuttedTexture.SetData(data);
                return cuttedTexture;
            }

and the methods ended too fast at the ‘//' ,leave the last two path no work.

**Comment by trinith** [comment]
I just ran a test to see if I could reproduce any issues and this worked for me...

```C#
// Generate a texture...
Texture2D texture = new Texture2D(_graphics.GraphicsDevice, 100, 100);
Color[] data = new Color[texture.Width * texture.Height];
for (int x = 0; x < texture.Width; x++)
{
    for (int y = 0; y < texture.Height; y++)
    {
        Vector2 p = new Vector2(x, y);
        p = new Vector2(50, 50) - p;
        float cValue = 1f - (p.Length() / 50f);
        data[texture.Width * x + y] = new Color(cValue, 0f, 0f);
    }
}
texture.SetData<Color>(data);
_spriteBatch.Draw(texture, new Vector2(10, 60), Color.White);

// Generate a clipped texture
Rectangle clipBounds = new Rectangle(25, 25, 50, 50);
Texture2D clippedTexture = new Texture2D(texture.GraphicsDevice, clipBounds.Width, clipBounds.Height);
Color[] clippedData = new Color[clipBounds.Width * clipBounds.Height];
texture.GetData<Color>(0, clipBounds, clippedData, 0, clippedData.Length);
clippedTexture.SetData<Color>(clippedData);
_spriteBatch.Draw(clippedTexture, new Vector2(10, 210), Color.White);
```

I'm having trouble understanding what you wrote but it sounds like maybe your program is crashing at your `texture.GetData(...)` call, which makes me wonder if your input texture might have a problem? I'm not sure, give my code a try. You can drop it right into a fresh project's draw call (assuming `_spriteBatch` exists) and make sure to put a Begin/End call around the code on your sprite batch.

It's also worth noting though that you don't have to manually slice up your textures in this way if you don't want to. There's an overload on the `Draw` method of `SpriteBatch` that will do this for you. For the code I posted above, the equivalent is...

```C#
_spriteBatch.Draw(texture, new Rectangle(10, 310, clipBounds.Width, clipBounds.Height), clipBounds, Color.White);
```

**Reply by Winlchance** [reply]
i found the problem. My visual studio don't update my code timely. so it run old code but display new code. the problem solve when i delete and add the project again.

i use manually cut instead spriteBatch.Draw because i just want get a cutted texture, and make code in Draw() more simple. thanks for answer.

**Reply by trinith** [reply]
> i use manually cut instead spriteBatch.Draw because i just want get a cutted texture, and make code in Draw() more simple. thanks for answer.

Ultimately I think it's best to implement your code in a way that makes sense to you. For most applications this probably won't make a difference; however, just so you have all the information I'll leave this comment here.

I believe there is an overhead cost every time the graphics API needs to switch textures. This is one of the main reasons folks use a texture atlas to begin with, so you can just load that texture into memory and do a lot of draws from it. There are also limits to how big a texture you want to load, so that's also why folks tend to organize similar things onto a texture atlas, such as all the animations for a particular character, or all of the map tiles for a particular level.

Slicing up a texture that represents a texture atlas into individual texture frames for all of your drawing will have a performance impact, but unless your game has high performance requirements, it probably won't matter much.

It's been a while since I dug into this stuff and it could just very well not matter anymore on modern hardware, so feel free to fact check me on this. As I said, it's your code and I'm not going to tell you how you have to implement, I just want to give you helpful information so you can make the best decision for you :)

Anyway, glad you foudn the problem! ❤️

**Comment by stromkos** [comment]
Just to add emphasis to a common rule:

"Do not modify a texture at runtime unless the changes are user in-game or game directed" 

"Save changes to variables when they happen, if needed."

`Texture.GetData()` and `Texture.SetData()` should only be called in `Update()` and should be considered slow operations, due to the sync wait on some platforms.

The "proper" way is to use a texture atlas is via source rectangles in your draw calls since a `Rectangle` uses a fixed amount of memory, 128 bytes used during draw (modified from defaults or not, it is applied into the draw transform matrix):
A computationally free operation.

The way to accomplish one-time singular modifications is usually done as a single step on install or config change,  that splits/modifies/combines/saves the image.

> make code in Draw() more simple. 

Code simplicity over performance should **never** be a reason in a production game.

Override the Draw() function to simplify calls if needed at runtime.
